R/utilities-mass-balance.R

```diff
@@ -357,3 +318,162 @@ getMassBalanceDataMapping <- function(plotType) {
+  # If groupings is NULL, use default inclusion/exclusion and grouping
+  groupings <- groupings %||% defaultMassBalanceGroupings(simulationResults, compoundNames)
+
+  massBalanceData <- data.frame()
+  previouslyIncludedMoleculePaths <- NULL
+  for (group in groupings) {
+    includedMolecules <- ospsuite::getAllMoleculesMatching(group$Include, simulation)
+    includedMoleculePaths <- sapply(includedMolecules, function(molecule) {
+      molecule$path
+    })
+    # Exclusion criteria
+    excludedMoleculePaths <- NULL
+    # If ExcludePreviousGroupings is TRUE, exclude molecules from previous groupings
+    if(group$ExcludePreviousGroupings %||% TRUE){
+      excludedMoleculePaths <- previouslyIncludedMoleculePaths
+    }
+    if (!is.null(group$Exclude)) {
+      excludedMolecules <- ospsuite::getAllMoleculesMatching(group$Exclude, simulation)
+      excludedMoleculePaths <- c(
+        excludedMoleculePaths,
+        sapply(excludedMolecules, function(molecule){molecule$path})
+      )
+    }
+    includedMolecules <- includedMolecules[!includedMoleculePaths %in% excludedMoleculePaths]
+    includedMoleculePaths <- setdiff(includedMoleculePaths, excludedMoleculePaths)
+    
+    if(isEmpty(includedMoleculePaths)) {
+      next
+    }
+    validateMoleculesFromCompounds(includedMolecules, compoundNames)
+    checkMoleculesAlreadyIncluded(includedMoleculePaths, previouslyIncludedMoleculePaths)
+    
+    previouslyIncludedMoleculePaths <- c(
+      previouslyIncludedMoleculePaths,
+      includedMoleculePaths
+    )
+
+    # For included molecule paths, get their amount vs time and sum them
+    moleculeAmounts <- sapply(
+      includedMoleculePaths,
+      function(moleculePath) {
+        simulationResults$getValuesByPath(moleculePath, individualIds = simulationResults$allIndividualIds)
+      }
+    )
+    moleculeAmounts <- rowSums(as.data.frame(moleculeAmounts))
+    
+    massBalanceData <- rbind.data.frame(
+      massBalanceData,
+      data.frame(
+        Time = simulationResults$timeValues,
+        Amount = moleculeAmounts,
+        Legend = group$Name,
+        stringsAsFactors = FALSE
+      )
+    )
+  }
+  return(massBalanceData)
+}
+
+#' @title defaultMassBalanceGrouping
+#' @description Get mass balance default inclusion/exclusion defined as groupings list
+#' In the Matlab version, the grouping was as followed:
+#' 1) Saliva is excluded from the mass balance calculation
+#' 2) Lumen is a parent compartment that regroups many sub-compartments but Feces
+#' 3) All the rest of the sub-compartments are used normally
+#' @param simulationResults A `SimulationResults` object
+#' @param compoundNames Names of simulation molecules
+#' @import ospsuite
+#' @keywords internal
+defaultMassBalanceGroupings <- function(simulationResults, compoundNames) {
+  groupNames <- unique(sapply(
+    simulationResults$allQuantityPaths,
+    function(path) {
+      head(tail(ospsuite::toPathArray(path), 2), 1)
+    }
+  ))
```

Comment:
seems wrong to me.
This should return something like {"Interstitial", "Intracellular", "Plasma", ...}, correct? But it returns the name of a compound instead.

```
> unique(sapply(
+     simulationResults$allQuantityPaths,
+     function(path) {
+       head(tail(ospsuite::toPathArray(path), 2), 1)
+     }
+   ))
[1] "Alfentanil"
```

Reply:
There should be a step before that only include molecule paths in the simulation results.
In the matlab code, the compartment to display in such molecule paths were compartment of the before the last element of the path array. 
Is it not actually the case ?

Reply:
ok, forget it, you are right.
In my test example I had a concentration instead of amount, that's why the returned grouping was wrong